tests/MongoDB.Driver.Tests/UnifiedTestOperations/UnifiedGridFsUploadOperation.cs

```diff
@@ -0,0 +1,118 @@
+
+namespace MongoDB.Driver.Tests.UnifiedTestOperations
+{
+    public class UnifiedGridFsUploadOperation : IUnifiedEntityTestOperation
+    {
+        private readonly IGridFSBucket _bucket;
+        private readonly string _filename;
+        private readonly GridFSUploadOptions _options;
+        private readonly byte[] _source;
+
+        public UnifiedGridFsUploadOperation(
+            IGridFSBucket bucket,
+            string filename,
+            byte[] source,
+            GridFSUploadOptions options)
+        {
+            _bucket = bucket;
+            _filename = filename;
+            _source = source;
+            _options = options;
+        }
+
+        public OperationResult Execute(CancellationToken cancellationToken)
+        {
+            try
+            {
+                var result = _bucket.UploadFromBytes(_filename, _source, _options, cancellationToken);
+
+                return OperationResult.FromResult(result);
+            }
+            catch (Exception exception)
+            {
+                return OperationResult.FromException(exception);
+            }
+        }
+
+        public async Task<OperationResult> ExecuteAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                var result = await _bucket.UploadFromBytesAsync(_filename, _source, _options, cancellationToken);
+
+                return OperationResult.FromResult(result);
+            }
+            catch (Exception exception)
+            {
+                return OperationResult.FromException(exception);
+            }
+        }
+    }
+
+    public class UnifiedGridFsUploadOperationBuilder
+    {
+        private readonly EntityMap _entityMap;
+
+        public UnifiedGridFsUploadOperationBuilder(EntityMap entityMap)
+        {
+            _entityMap = entityMap;
+        }
+
+        public UnifiedGridFsUploadOperation Build(string targetBucketId, BsonDocument arguments)
+        {
+            var bucket = _entityMap.GetBucket(targetBucketId);
+
+            string filename = null;
+            GridFSUploadOptions options = null;
+            byte[] source = null;
+
+            foreach (var argument in arguments)
+            {
+                switch (argument.Name)
+                {
+                    case "chunkSizeBytes":
+                        options = options ?? new GridFSUploadOptions();
+                        options.ChunkSizeBytes = argument.Value.AsInt32;
+                        break;
+                    case "filename":
+                        filename = argument.Value.AsString;
+                        break;
+                    case "source":
+                        var sourceDocument = argument.Value.AsBsonDocument;
+                        JsonDrivenHelper.EnsureAllFieldsAreValid(sourceDocument, "$$hexBytes");
+                        var sourceString = sourceDocument["$$hexBytes"].AsString;
+                        (sourceString.Length % 2).Should().Be(0);
```

Comment:
Don't use `Should` here. Replace with:

    if (sourceString.Length %2 != 0)
    {
        throw new FormatException("$$hexBytes must have an even number of bytes.");
    }

Reply:
Or... remove the check entirely. `BsonUtils.ParseHexString` doesn't care if the length is odd (because a leading `0` can be omitted).